repository/src/interpret/works.rs

```diff
@@ -0,0 +1,337 @@
+            Err(CommitError::Commit(error, commit)) => {
+                return Ok(Err(format!("failed to parse commit {commit}: {error}",)));
+            }
+            Err(e) => return Err(e.into()),
+        };
+        for (commit, commit_hash) in &commits {
+            if let Err(e) = csv.apply_commit(commit) {
+                return Ok(Err(format!(
+                    "commit sequence verification failed: {e} at {commit_hash}",
+                )));
+            }
+        }
+
+        // If the commit sequence contains block commit(s) that can be finalized
+        let headers = csv.get_block_headers();
+        if headers.len() > 2 {
+            let (last_header, _) = headers.last().expect(
+                "already checked that the received commit is not same as the last finalized block",
+            );
+            let (second_to_last_header, index) = headers[headers.len() - 2].clone();
+            advance_finalized_branch(
+                raw,
+                commits[index].1,
+                LastFinalizationProof {
+                    height: second_to_last_header.height,
+                    proof: last_header.prev_block_finalization_proof.clone(),
+                },
+            )
+            .await?;
+        }
+
+        // Create the associate branch
+        let branch_name = match &commits
+            .last()
+            .expect("already checked that the received commit is not a finalization proof commit")
+            .0
+        {
+            Commit::Agenda(agenda) => {
+                format!(
+                    "a-{}",
+                    &agenda.to_hash256().to_string()[0..BRANCH_NAME_HASH_DIGITS]
+                )
+            }
+            Commit::AgendaProof(agenda_proof) => {
+                format!(
+                    "a-{}",
+                    &agenda_proof.to_hash256().to_string()[0..BRANCH_NAME_HASH_DIGITS]
+                )
+            }
+            Commit::Block(block) => {
+                format!(
+                    "b-{}",
+                    &block.to_hash256().to_string()[0..BRANCH_NAME_HASH_DIGITS]
+                )
+            }
+            x => return Ok(Err(format!("commit sequence ends with: {x:?}"))),
+        };
+        if raw.locate_branch(branch_name.clone()).await.is_ok() {
+            return Ok(Err(format!("branch already exists: {branch_name}",)));
+        }
+        raw.create_branch(branch_name, tip_commit_hash).await?;
+    };
+    Ok(Ok(()))
+}
+
+pub async fn sync_all(raw: &mut RawRepository) -> Result<Vec<(String, Result<(), String>)>, Error> {
+    let local_branches: Vec<String> = raw
+        .list_branches()
+        .await?
+        .into_iter()
+        .filter(|s| {
+            s.as_str() != WORK_BRANCH_NAME
+                && s.as_str() != FINALIZED_BRANCH_NAME
+                && s.as_str() != FP_BRANCH_NAME
+                && !s.starts_with("a-")
+                && !s.starts_with("b-")
+        })
+        .collect();
```

Comment:
Was only leaving local p branches your intention?

Reply:
Nope mistake. fixed.

Reply:
Btw I believe `p` is more appropriate as a tag. Let's revisit this later